stablehlo/reference/InterpreterApi.cpp

```diff
@@ -44,47 +43,63 @@ func::FuncOp getMainFunction(ModuleOp module, StringRef mainName) {
 
   return {};
 }
-}  // namespace
 
-llvm::Error InterpreterFallback::operator()(
-    const InterpreterConfiguration &config, Operation &op, Scope &scope,
-    Process *process) {
-  llvm::StringRef funcName = op.getParentOfType<func::FuncOp>().getSymName();
-
-  if (auto probeOp = dyn_cast<stablehlo::interpreter::ProbeOp>(op)) {
-    auto input =
-        stablehlo::InterpreterValue(scope.findTensor(probeOp.getOperand()));
-    auto status = stablehlo::interpreter::evalProbeOp(
-        input, probeOp.getProbeId(), config.probeInstrumentationDir,
-        instrumentedTensors);
-    scope.add(probeOp.getResult(), input);
-    return wrapFallbackStatus(std::move(status), funcName, "interpreter.probe");
+// Register a default fallback callback to handle select interpreter ops.
+class DefaultInterpreterFallback : public InterpreterFallback {
+ public:
+  DefaultInterpreterFallback(const InterpreterConfiguration &config)
+      : config(config){};
+
+  virtual llvm::Error operator()(Operation &op, Scope &scope,
+                                 Process *process) final {
+    llvm::StringRef funcName = op.getParentOfType<func::FuncOp>().getSymName();
+
+    if (auto probeOp = dyn_cast<stablehlo::interpreter::ProbeOp>(op)) {
+      auto input =
+          stablehlo::InterpreterValue(scope.findTensor(probeOp.getOperand()));
+      auto status = stablehlo::interpreter::evalProbeOp(
+          input, probeOp.getProbeId(), config.probeInstrumentationDir,
+          instrumentedTensors);
+      scope.add(probeOp.getResult(), input);
+      return wrapFallbackStatus(std::move(status), funcName,
+                                "interpreter.probe");
+    }
+
+    if (auto runParallelOp =
+            dyn_cast<stablehlo::interpreter::RunParallelOp>(op)) {
+      auto runtimeOperands = scope.find(runParallelOp.getInputs());
+      std::queue<StringAttr> infeed;
+      if (auto infeedAttr = runParallelOp.getInfeed())
+        for (auto &value : infeedAttr->getValue())
+          infeed.push(value.cast<FlatSymbolRefAttr>().getAttr());
+
+      SmallVector<SmallVector<StringAttr>> programs(
+          runParallelOp.getPrograms().size());
+      for (auto [i, replica] : llvm::enumerate(runParallelOp.getPrograms()))
+        for (auto &program : replica.cast<ArrayAttr>())
+          programs[i].push_back(program.cast<FlatSymbolRefAttr>().getAttr());
+
+      SymbolTable symbolTable{op.getParentOfType<ModuleOp>()};
+      auto results = stablehlo::interpreter::evalRunParallelOp(
+          runtimeOperands, infeed, programs, symbolTable);
+      scope.add(runParallelOp.getResults(), results);
+      return wrapFallbackStatus(llvm::Error::success(), funcName,
+                                "interpreter.run_parallel");
+    }
+
+    return (*config.fallback)(op, scope, process);
```

Comment:
If config points back to this class, which I think is a typical usage(?) wouldn't this recurse infinitely?

Reply:
Ah. This is an internal implementation detail. Please add a comment to the class saying "DefaultInterpreterFallback is an implementation detail of run module. It takes in an InterpreterConfiguration which can have user-implemented fallbacks"

As of now, this setup means that a custom op cannot have a region that uses a default callback (i.e. my.op with a region containing a probe op). This is probably an OK limitation for now, but would require refactoring into something more composable if this ever becomes a requirement.